statsmodels/kernel_methods/_kde1d_reflection.py

```diff
@@ -0,0 +1,239 @@
+            terms -= kernel.cdf((L - exog) / bw)  # Remove the truncated part on the left
+            terms += kernel.cdf(z1 - (2 * L / bw))  # Add the reflected part
+            terms -= kernel.cdf((exog - L) / bw)  # Remove the truncated part from the reflection
+
+        if U < np.inf:
+            terms += kernel.cdf(z1 - (2 * U / bw))  # Add the reflected part
+
+        terms *= self.weights
+        terms.sum(axis=-1, out=out)
+        out /= self.total_weights
+
+        out[points[..., 0] >= self.upper] = 1
+        out[points[..., 0] <= self.lower] = 0
+
+        return out
+
+    def grid(self, N=None, cut=None, span=None):
+        """
+        DCT-based estimation of KDE estimation, i.e. with reflection boundary
+        conditions. This works only for fixed bandwidth (i.e. adjust = 1) and
+        gaussian kernel.
+
+        For open domains, the grid is taken with 3 times the bandwidth as extra
+        space to remove the boundary problems.
+        """
+        if self.adjust.shape:
+            return KDE1DMethod.grid(self, N, cut)
+
+        bw = self.bandwidth * self.adjust
+        exog = self.exog
+        N = self.grid_size(N)
+
+        if span is None:
+            lower = self.lower
+            upper = self.upper
+        else:
+            lower, upper = span
+
+        if cut is None:
+            cut = self.kernel.cut
+
+        if not finite(lower):
+            lower = np.min(exog) - cut * self.bandwidth
+        if not finite(upper):
+            upper = np.max(exog) + cut * self.bandwidth
+
+        weights = self.weights
+
+        return dctdensity(exog, self.kernel.dct, bw, lower, upper, N, weights, self.total_weights)
+
+    def from_binned(self, mesh, binned, normed=False, dim=-1):
+        """
+        Evaluate the PDF from data already binned. The binning might have been high-dimensional but must be of the same
+        data.
+
+        Parameters
+        ----------
+        mesh: grid.Grid
+            Grid of the binning
+        bins: ndarray
+            Array of the same shape as the mesh with the values per bin
+        normed: bool
+            If true, the result will be normed w.r.t. the total weight of the exog
+        dim: int
+            Dimension along which the estimation must be done
+
+        Returns
+        -------
+        ndarray
+            Array of same size as bins, but with the estimated of the PDF for each line along the dimension `dim`
+        """
+        if self.adjust.ndim:
+            raise ValueError("Error, cannot use binned data with non-constant adjustment.")
+        return dctdensity_from_binned(mesh, binned, self.kernel.dct, self.bandwidth*self.adjust,
+                                      normed, self.total_weights, dim=dim)
+
+    def grid_size(self, N=None):
+        """
+        Returns a valid grid size.
+        """
+        if N is None:
+            if self.adjust.shape:
+                return 2 ** 10
+            return 2 ** 16
```

Comment:
I chose 2**16 because it's was fast enough on most computers I use, it's not too much memory, and accuracy depends on the number of slots used for binned methods. That being said, the main reason why `grid_size` is a method and not a function is to allow for more complex behaviour if we need. For example, we could test the number of samples to return something related to it.